xarray/core/dataset.py

```diff
@@ -3867,6 +3867,79 @@ def differentiate(self, coord, edge_order=1, datetime_unit=None):
                 variables[k] = v
         return self._replace_vars_and_dims(variables)
 
+    def integrate(self, dim, datetime_unit=None):
+        """ integrate the array with the trapezoidal rule.
+
+        .. note::
+            This feature is limited to simple cartesian geometry, i.e. coord
+            must be one dimensional.
+
+        Parameters
+        ----------
+        dim: str, or a sequence of str
+            Coordinate(s) used for the integration.
+        datetime_unit
+            Can be specify the unit if datetime coordinate is used. One of
+            {'Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms', 'us', 'ns', 'ps', 'fs',
+             'as'}
+
+        Returns
+        -------
+        integrated: Dataset
+
+        See also
+        --------
+        DataArray.integrate
+        numpy.trapz: corresponding numpy function
+        """
+        if not isinstance(dim, (list, tuple)):
+            dim = (dim, )
+        result = self
+        for d in dim:
+            result = result._integrate_one(d, datetime_unit=datetime_unit)
+        return result
+
+    def _integrate_one(self, dim, datetime_unit=None):
+        from .variable import Variable
+
+        if dim not in self.variables and dim not in self.dims:
+            raise ValueError('Coordinate {} does not exist.'.format(dim))
+
+        coord_var = self[dim].variable
+        if coord_var.ndim != 1:
```

Comment:
I think this is currently not possible due to xarray's data model, but it's a good idea to add this anyways given that we want to change this soon (e.g., see https://github.com/pydata/xarray/pull/2405).

I would recommend adjusting this to `if coord_var.dims != (dim,)`, which is a little stricter.

Reply:
I first thought that it would be nice if we could integrate even along non-dimensional (1d) coordinate (as `interpolate_na`, `differential` do), but it also sounds something too much.
How do you think?

Reply:
Then, `coord` is a better argument rather than `dim`? 
Or we use `dim` for argument but support integration along non-dimensional coordinate with a slight avoidance of correctness, as it is more consistent with other reduction methods?

Reply:
I don't have a strong opinion here.

Reply:
Well, `differentiate` uses `coord`, so maybe `integrate` should too?